### ISSUE TITLE
[HOLD for payment 2024-11-11] [HOLD for payment 2024-11-05] [$250] De-dupe AddComment, EditComment, and DeleteComment requests

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#50074]
### Problem

When we make unnecessary network requests, it slows down the app and contributes to higher traffic on our servers. In particular:

- if you are offline and you serialize multiple `EditComment` requests, only the last one will take effect
- if you serialize both an `AddComment` and any number of `EditComment` requests, the result would be the same as one `AddComment` with the content of the final `EditComment`
- if you serialize an `AddComment` followed by a `DeleteComment`, the end result is the same as not sending any network requests at all
- if you serialize an `EditComment` request followed by a `DeleteComment`, the end result is the same as only sending the `DeleteComment`

### Solution

De-dupe the network requests!
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021848400026446902359</li>
        <li>Upwork Job ID: 1848400026446902359</li>
        <li>Last Price Increase: 2024-10-21</li>
<li>Automatic offers: </li>
<ul>
<li>dukenv0307 | Reviewer | 104528746</li>
<li>c3024 | Contributor | 104528756</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @bfitzexpensify</details>

**Comment by gedu** [comment]
Hey, I'm Edu from Callstack, I can take this one

**Comment by roryabraham** [comment]
> The obstacle here is that if the request fails after replace call, it will show wrong comment.

isn't that the case today with any `EditComment` request? Or is there something I'm missing?

**Comment by roryabraham** [comment]
Going on parental leave. @gedu @mountiny @shubham1206agra I leave this in your capable hands. My parting note is to please use as many automated tests to cover this change as possible 👍🏼

**Comment by gedu** [comment]
@shubham1206agra I'm making me a diagram, and I was thinking, what about a reaction, `deleteComment` has to delete `addEmojiReaction` too. Because now if you edit -> edit -> add reaction -> delete, you will see a request for each in that order. When we should see only a delete request, right?

**Comment by gedu** [comment]
@shubham1206agra also just to make sure, if I edit -> create a thread -> add message in that thread -> delete original message, this is the final state. In such case I only delete the `editComment` the thread should be created and any message in it, right?


<img width="965" alt="Screenshot 2024-10-07 at 18 46 36" src="https://github.com/user-attachments/assets/69b3d5b9-4557-4c64-a5e6-2b26ad8fa935">
This is how it looks now

**Comment by gedu** [comment]
> addEmojiReaction

Ohh, yeah, so nothing happens, no new thread is created and no new message is added into that thread. Cool, I'm adding this case too

**Comment by gedu** [comment]
Just to make sure, these cases can happen at any time, not just offline. Given that the queue is being processed sequentially, the delete could occur at any time, and I have to prevent those messages from being sent, right?

**Comment by gedu** [comment]
> Can you report the bug in slack?

How can I do it? I don't have access to the prompt /bug

**Comment by mountiny** [comment]
@gedu are you able to start a PR for this one? anyone from callstack team able to help?

**Comment by gedu** [comment]
Yes, I'm working on this, started already with the Delete comments

**Comment by gedu** [comment]
what should be the best approach to this case:
```
Offline -> add message -> update -> delete message (now what happens is the message is strikethrough) because when is 
```
back online after the response from the server it is removed from Onyx.
But now because I won't do any request to the server I need to clean it, I can just use the `successData` but when?

1) as soon as the user deletes the message?
2) Should I wait until the user is back online (this is just for the user to see the message strikethrough)

2. will require more logic and can make everything a bit more complicated

Note: this is `only` when a message  is created  while the user is offline, if the message already exist I need to send the DeleteMessage request

personally I vote for `1`

@mountiny @shubham1206agra

**Comment by gedu** [comment]
I'm removing the message also from Onyx, and I'm writing the test, still working on the case:
```
Offline -> add message -> update -> delete message
```
because it is the more complex one

**Comment by mountiny** [comment]
Yeah I think 1 makes sense. I think all of these flows will have to have thorough unit tests. This is super easy to mess up so we need to cover it up with automation.

**Comment by gedu** [comment]
The code is working. I'm adding tests for more cases.

when adding a comment with attachment, just attachment, with emoji reactions (adding and removing)
Also testing those cases when the deleteComment has to be sent to backend and when no.

**Comment by gedu** [comment]
I'm handling the case where a comment is used to create a thread and then gets deleted. In this situation, I'm not removing it from the queue because I don't know if there will be more messages or deletions. Additionally, it seems that an OpenReport must be done to create the thread, and I can't delete the comment because it will fail. Now I will be working on the tests

**Comment by gedu** [comment]
@mountiny I'm still working on the DeleteComment, and I'm asking @zirgule to help me with EditComment since it can be solved in parallel.

**Comment by zirgulis** [comment]
hi @gedu @mountiny, sure I will take the EditComment part, will post my update here tomorrow.

**Comment by gedu** [comment]
@mountiny @shubham1206agra created a PR for the DeleteComment, so we can start reviewing it. https://github.com/Expensify/App/pull/50919

**Comment by gedu** [comment]
> if you serialize both an AddComment and any number of EditComment requests, the result would be the same as one 
> AddComment with the content of the final EditComment

After a talk with @zirgulis that point we will tackle in a new PR, not in the same one of multiple `EditComment`, given that it will require more logic and a new set of tests. That way we can track it better if any issue shows up

**Comment by zirgulis** [comment]
> if you are offline and you serialize multiple EditComment requests, only the last one will take effect

@mountiny @gedu for this I have the code ready [here](https://github.com/zirgulis/expensify/commit/e42d272bff0734442e2d4654c6de83c87587aaab), will create a PR tomorrow since I still need to test on emulators and do the screen recordings.

**Comment by zirgulis** [comment]
@mountiny  @gedu  Multiple EditComment requests deduping PR is open https://github.com/Expensify/App/pull/51149

**Comment by mountiny** [comment]
Thanks!

**Comment by gedu** [comment]
> if you serialize both an AddComment and any number of EditComment requests, the result would be the same as one > AddComment with the content of the final EditComment

@mountiny I did the second item, and I'm seeing this behaviour, for a couple of seconds, until the response is back and then it rerenders, you can see the `edit` label next to the message. Is it ok?

Given that the `optimisticData` is applied before the request reaches the SequentialQueue, it is showing, I can try to add a rollback case for it, but want to confirm if we want or not to show the `edit` label. The rollback will be applied as soon as the "conflict" is resolved so probably the user won't see the `edit` label while being of line, will be like he just updated the comment.

**Comment by mountiny** [comment]
@gedu I think from UX it would be best not to show it but I would not block on this if its too complex to achieve not showing edited label

**Comment by gedu** [comment]
> if you serialize both an AddComment and any number of EditComment requests, the result would be the same as one > > AddComment with the content of the final EditComment

Created a new PR: https://github.com/Expensify/App/pull/51422 to add the last point, it depends on https://github.com/Expensify/App/pull/50919, so would be good to merge that one first

**Comment by mountiny** [comment]
2 PRs been tested (https://github.com/Expensify/App/pull/50919, https://github.com/Expensify/App/pull/51422) and we merged the one that had more stuff in it. As a result we will pay $250 to both @c3024 and to @dukenv0307

**Comment by bfitzexpensify** [comment]
Thanks all - payments complete.
